log4j-api/src/main/java/org/apache/logging/log4j/spi/internal/DefaultScopedContextProvider.java

```diff
@@ -0,0 +1,421 @@
+import org.apache.logging.log4j.Logger;
+import org.apache.logging.log4j.ScopedContext;
+import org.apache.logging.log4j.ThreadContext;
+import org.apache.logging.log4j.spi.ScopedContextProvider;
+import org.apache.logging.log4j.status.StatusLogger;
+import org.apache.logging.log4j.util.StringMap;
+
+/**
+ * An implementation of {@link ScopedContextProvider} that uses the simplest implementation.
+ * @since 2.24.0
+ */
+public class DefaultScopedContextProvider implements ScopedContextProvider {
+
+    public static final Logger LOGGER = StatusLogger.getLogger();
+    public static final ScopedContextProvider INSTANCE = new DefaultScopedContextProvider();
+
+    private final ThreadLocal<Instance> scopedContext = new ThreadLocal<>();
+
+    /**
+     * Returns an immutable Map containing all the key/value pairs as Object objects.
+     * @return The current context Instance.
+     */
+    private Optional<Instance> getContext() {
+        return Optional.ofNullable(scopedContext.get());
+    }
+
+    /**
+     * Add the ScopeContext.
+     * @param context The ScopeContext.
+     */
+    private void addScopedContext(final Instance context) {
+        Instance current = scopedContext.get();
+        scopedContext.set(context);
+    }
+
+    /**
+     * Remove the top ScopeContext.
+     */
+    private void removeScopedContext() {
+        final Instance current = scopedContext.get();
+        if (current != null) {
+            if (current.parent != null) {
+                scopedContext.set(current.parent);
+            } else {
+                scopedContext.remove();
+            }
+        }
+    }
+
+    @Override
+    public Map<String, ?> getContextMap() {
+        final Optional<Instance> context = getContext();
+        return context.isPresent()
+                        && context.get().contextMap != null
+                        && !context.get().contextMap.isEmpty()
+                ? Collections.unmodifiableMap(context.get().contextMap)
+                : Collections.emptyMap();
+    }
+
+    /**
+     * Return the value of the key from the current ScopedContext, if there is one and the key exists.
+     * @param key The key.
+     * @return The value of the key in the current ScopedContext.
+     */
+    @Override
+    public Object getValue(final String key) {
+        final Optional<Instance> context = getContext();
+        return context.map(instance -> instance.contextMap)
+                .map(map -> map.get(key))
+                .orElse(null);
+    }
+
+    /**
+     * Return String value of the key from the current ScopedContext, if there is one and the key exists.
+     * @param key The key.
+     * @return The value of the key in the current ScopedContext.
+     */
+    @Override
+    public String getString(final String key) {
+        final Optional<Instance> context = getContext();
+        if (context.isPresent()) {
+            final Object obj = context.get().contextMap.get(key);
```

Comment:
Here you (correctly) assume that `contextMap` is not null, so there will be no NPE.

Reply:
This is no longer applicaable.